### DISCUSSION TITLE
[Niche case support] keeping 2 latest versions (latest & latest-1) in my package json file

### DISCUSSION THREAD

**Discussion by Devinwong** [renovatebot/renovate#31057] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

Hi Renovate community.
This is probably a niche case but I still wanted to check if there happens to have any smart way to achieve this request.
The question I have is, is there a way for Renovate to monitor and keep the latest 2 versions (latest & latest-1) in my package json file? For some reasons, we need both versions to support some compatibility scenarios.
For example, my package json file looks like
```
{
  "packageName": "package1",
  "versions": [
    {"version": "1.0.1"},
    {"version": "1.0.0"}
}
```
When Renovate detects there is a latest version `1.0.2`. The PR will become
```
{
  "packageName": "package1",
  "versions": [
    {"version": "1.0.2"},
    {"version": "1.0.1"}
}
```
where it always keeps the latest 2 versions. Is this something feasible?

If I break it down into 2 steps. They will be
1. Update "1.0.1" to latest (1.0.2), which is the default behavior of Renovate. So this is good.
2. Update "1.0.0" to second latest (1.0.1). Not sure if Renovate has the concept of second latest?

The key name `version`s can be changed to `latestVersion` and `secondLatestVersion` if needed.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by RahulGautamSingh** [comment, marked as answer]
No its not possible.

But if you are open for a hacky one, [here's something](https://github.com/Rahul-renovate-testing/second-latest) I tried as I thought `autoReplaceStringTemplate` could get this done.

Btw, its not strictly `latest` and `latest-1` instead it updates the first version with latest version and for the second version it just updates it with the value of first version. Also, the formatting is not perfect so make sure to correct that afterwards.

**Reply by Devinwong** [reply]
Wow this is awesome. Thanks for sharing an example too! It could save me a few days.
Let me take a closer look at get back to this post later. Appreciate your help!

**Reply by Devinwong** [reply]
@RahulGautamSingh , I tried to follow your example to do the same thing in my minimal reproduction but failed with error
```
DEBUG: Could not extract test_components_simple.json (manager=regex) after autoreplace. Did the autoreplace make the file unparseable? (branch="renovate/moby-containerd-1.x")
WARN: Error updating branch: update failure (branch="renovate/moby-containerd-1.x")
```

This is my repro. https://github.com/Devinwong/RenovateTest/tree/main
For more context, if I remove line 38 `autoReplaceStringTemplate` in renovate.json, it could create a PR as expected.
But if I add that `autoReplaceStringTemplate`, it just reported the error message as above. Do you know how I can get more details about why it failed? Like the output of the autoreplace.

Thanks for your time.

**Reply by RahulGautamSingh** [reply]
No need to escape the period in ubuntu version with double slashes in the `autoReplaceStringTemplate`.

Try this:

``` 
"autoReplaceStringTemplate": "\"renovateTag\": \"name={{{packageName}}} os=ubuntu release=22.04\",\n\"latestVersion\": \"{{{newVersion}}}\",\n\"secondLatestVersion\": \"{{{currentValue}}}\"" 
```

**Reply by Devinwong** [reply]
Oh right! That escape was for regex. A general string escape doesn't need double slashes. It works now. Thanks for this hacky but super helpful way!
